### PR TITLE
Perf/pool trie store

### PR DESCRIPTION
**Pull request by asdacap**
- Remove commit lock in snap sync.

## Changes

- Remove commit lock in snap sync by using some pool of triestore.
- Seems increase the upper bound of rate of state sync on my machine from 15Mil to 20Mil. But its generally inconsistent as always with snap sync due to its need to get good peers as it starts early.

(yellow is the lock, green is commit, the 5th, 6th and 7th run have a special rocksdb tune to reduce write amp)
Screenshot from 2023-03-09 10-58-08

#### What types of changes does your code introduce?

- [X] Optimization

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [X] No

#### Notes on testing

- Mainnet can sync. Seems to improve performance by about 20 to 25%. 
- Goerli can sync. Peak writes rate with goerli is higher at 55Mil node per sec. Finished with a single pivot.
- Sepolia can sync, after a long wait of finding peer.

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs [hunk 1]
```diff
@@ -97,6 +95,7 @@ public AddRangeResult AddAccountRange(long blockNumber, Keccak expectedRootHash,
                 _logger.Trace($"SNAP - AddAccountRange failed, expected {blockNumber}:{expectedRootHash} but was {tree.RootHash}, startingHash:{startingHash}");
             }
 
+            _trieStorePool.Return(store);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs:98]: in finally clause?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs [hunk 2]
```diff
@@ -182,12 +182,14 @@ public AddRangeResult AddStorageRange(long blockNumber, PathWithAccount pathWith
                 _progressTracker.EnqueueAccountRefresh(pathWithAccount, startingHash);
             }
 
+            _trieStorePool.Return(store);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs:185]: same here

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs [hunk 3]
```diff
@@ -240,6 +242,7 @@ public void RefreshAccounts(AccountsToRefreshRequest request, byte[][] response)
             }
 
             _progressTracker.ReportAccountRefreshFinished();
+            _trieStorePool.Return(store);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs:245]: same here

**Review by LukaszRozmej — COMMENTED** [review]
Should those TrieStore's share `_objectsCache` ?

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs]: Done

**Comment by asdacap** [comment]
Fair point.

**Comment by asdacap** [comment]
Ow no, that is a part of dirty nodes. I would not take that risk.

**Review by LukaszRozmej — APPROVED** [review]
Looks good but please do rigorous testing including also sepolia.

**Comment by benaadams** [comment]
>  have a special rocksdb tune to reduce write amp

? 👀

**Comment by asdacap** [comment]
See branch `feature/sync-db-tuner`. I'll make a PR on it eventually....

**Comment by asdacap** [comment]
Made a PR on that #5410.

**Pull request merged**